FirebaseDynamicLinks/Sources/FIRDynamicLinks.m

```diff
@@ -440,6 +440,38 @@ - (nullable FIRDynamicLink *)dynamicLinkFromUniversalLinkURL:(NSURL *)url {
   return nil;
 }
 
+- (void)dynamicLinkFromUniversalLinkURL:(NSURL *)url
```

Comment:
dynamicLinkFromUniversalLinkURL:completion and dynamicLinkFromUniversalLinkURL both shares the same code apart from completion block calling part. So can we refactor the common code to a method which can be called from dynamicLinkFromUniversalLinkURL:completion and dynamicLinkFromUniversalLinkURL which would avoid code duplication and would make the code maintainable.

Reply:
`completion` callback is called in multiple (2) places inside `dynamicLinkFromUniversalLinkURL:completion` method. Moreover, these methods differ in return value - `void` vs `FIRDynamicLink`. The idea is to deprecate `dynamicLinkFromUniversalLinkURL:` now and remove completely in the upcoming releases. Do you still think it makes sense to refactor it? If you insist though, I can do it. No problem.

Reply:
I agree with the fact that the other function is getting deprecated and will get eventually removed but we are not sure when we are going to do it. Since removing the other method would be a breaking change, I'm not sure that we can do it sooner. So until then, if we need to make any fixes in dynamicLinkFromUniversalLinkURL logic, we will end up having it in two places.

My thought was to have a common internal method which can take the completion handler as well returns the dynamic link so that when called from `dynamicLinkFromUniversalLinkURL:` we will return based on the return of this internal method otherwise if called from `dynamicLinkFromUniversalLinkURL:completion`,we will ignore the return and rely on the completion block. So my proposal would be something like: 

    (nullable FIRDynamicLink *)dynamicLinkInternalFromUniversalLinkURL:(NSURL *)url
                                  completion:(FIRDynamicLinkUniversalLinkHandler)completion {
     //Have existing logic here and make call to completion block only if its not nil.
     ....
     if(completion){
     completion(dynamicLink,error);
     }
     ....

When calling this internal function from `dynamicLinkFromUniversalLinkURL:` just do :

    return  [self dynamicLinkInternalFromUniversalLinkURL:url completion:nil]

  And when calling from `dynamicLinkFromUniversalLinkURL:completion` just call : 

    [self dynamicLinkInternalFromUniversalLinkURL:url completion:completion]

I'm not sure this is a good coding pattern but this would help us with code repetition. Once we remove the old method completely, we can get rid of this internal method. Since my experience with Obj C is limited, I would prefer to get advice on this from @paulb777

Reply:
Agreed. It's worth refactoring to a common internal method - even if it's only a few months - since this much duplicate code could easily lead to inconsistent behavior or bugs.

Reply:
@palexsua could you please address this prior to merging?

Reply:
@eldhosembabu Addressed in https://github.com/firebase/firebase-ios-sdk/pull/5853/commits/0944814ddf4b11cf59f258439797ce3e8dd965a5

Reply:
thanks @palexsua